docs/flip/1009-optimizer-api.md

```diff
@@ -0,0 +1,541 @@
+      p = list(traversal.iterate(params))
+      g = list(traversal.iterate(grads))
+      new_p, new_state = opt.update(g, p, state)
+      new_params = traversal.set(new_p, new_params)
+      new_states.append(new_state)
+    return new_params, new_states
+
+  return Optimizer(init_fn, update_fn)
+```
+
+Remarks:
+
+- The `flax.optim.ModelParamTraversal` is reused.
+- The interface is a bit simplified, allowing to filter by path only (filtering
+  by value was never used), and shortening user code a bit by instantiating the
+  `ModelparamTraversal` inside the `multi_optimizer()` function.
+- We keep the same behavior as in the previous implementation, where parameters
+  are updated by the last optimizer in the sequence, in the case where a path
+  matches more than one `FilterFn`.
+- Note that the signature in `Optimizer` from the [functional API] needs to be
+  changed slightly to accept and return
+  `Union[OptimizerState, Sequence[OptimizerState]]` to make this work.
+
+## Functional Usage
+[Functional Usage]: #functional-usage
+
+Using the [functional API] directly makes for the following code:
+
+```python
+def make_train_step(apply_fn, opt_update_fn):
+  @jax.jit
+  def train_step(opt_state, variables, inputs, labels):
+
+    def loss_fn(params):
+      logits, new_model_state = apply_fn(
+          variables.copy(dict(params=params)), inputs, mutable=['batch_stats'])
+      loss = xent_loss(logits, labels)
+      return loss, new_model_state
+
+    variables, params = variables.pop('params')
+    (loss, new_model_state), grads = jax.value_and_grad(loss_fn, has_aux=True)(
+        params)
+    new_params, new_opt_state = opt_update_fn(
+        grads, params, opt_state)
+    new_variables = variables.copy(dict(params=new_params, **new_model_state))
+    return new_opt_state, new_variables, loss
+
+  return train_step
+
+opt_state = optim.init(variables['params'])
+train_step = make_train_step(model.apply, optim.update)
+for batch in ds.as_numpy_iterator():
+  opt_state, variables, loss = train_step(
+      opt_state, variables, batch['image'], batch['label'])
+```
+
+Remarks:
+
+- Compared with the [previous API], we can now keep the entire `variables`
+  including the `params` as an input and output to the `train_step()`.
+- Splitting `params` from `variables` is still necessary inside the train step
+  because we only want to compute gradients with respect to `params` and not the
+  entire `variables`.
+- A further complication is that we cannot provide `apply_fn` and
+  `opt_update_fn` as direct arguments to the jitted function `train_step()`
+  because they're not valid JAX types - they need to be wrapped in a
+  `make_train_step()` function like above, or declared as `static_argnums` to
+  `jax.jit()`.
+
+## Linen Helper
+[Linen helper]: #linen-helper
+
+While the above training loop is perfectly fine, a number of lines are going to
+be exactly the same for every Linen training loop, namely:
+
+1. The splitting of variables.
```

Comment:
I think many of these things indeed are boilerplate but not all of them. For example the assumption that there's a single `apply_fn` isn't accurate in many cases and I think that would be better off written explicitly.

The "variable collection" munging (your items 1-4 -- taking params out of variables and back in) is boilerplate for 99% of cases though and I agree that it's annoying now and we should improve on it.

So I'm wondering what the "Linen Helper" that only does 1-4 look like?

Another option, perhaps, is simply to stop using variable dictionaries as idiomatic Flax training code and instead keep track of "params" and "other_variables", e.g.

```
params = ...
other_variables = {'batch_stats': ..., ...}

output, other_variables = module.apply({"params": params, **other_variables}, inputs, mutable=['batch_stats'])
```

Reply:
I wonder if structuring the code like that reduces the boilerplate for optimization steps as well?

Reply:
I think it's also worth considering a very common use case is transfer learning, so users will want to freeze some subset of parameters and train others.

Reply:
okok maybe that went too far :) 

@avital reduced the pattern to a `TrainState` with some predefined fields and helper `.param` property & `.update()` function, PTAL

@n2cholas yeah that's a good point where you would want to have more control over how variables are split; I think in that case you could either do it manually in the training step or override `.param` & `.update()` accordingly (referring to the updated section `Train State` in the FLIP)

Reply:
Thanks @andsteing, that makes sense. Perhaps I'm missing something, so please correct me if I'm wrong, but what's the value in providing this `update()` method? Isn't the method equivalent to:

```python
state = state.replace(
    ...
    variables={**variables, 'params': new_params, **new_model_state}
)
```

Reply:
If providing support for easy transfer learning is out of scope for this proposal, you can disregard this comment. But if it is in scope, another potential idea is to make state, trainable params, and frozen params more explicit. This sort-of mirrors how Keras provides access to `trainable_weights` and `non_trainable_weights` separately.

```python
class TrainState(flax.struct.PyTreeNode):
  apply_fn: Callable = flax.struct.field(pytree_node=False)
  trainable_params: flax.core.FrozenDict[str, Any]
  frozen_params: flax.core.FrozenDict[str, Any]
  model_state: flax.core.FrozenDict[str, Any]
  tx: optax.GradientTransformation = flax.struct.field(pytree_node=False)
  opt_state: optax.OptState
        
  @property
  def params(self):
    return merge(self.trainable_params, self.frozen_params)
    
  @property
  def variables(self):
    return {'params': self.params, **self.model_state}
```

Where `merge` behaves like [`haiku.data_structures.merge`](https://dm-haiku.readthedocs.io/en/latest/api.html#haiku.data_structures.merge) (recursively merges sub-dicts as well). A rough implementation for this in pure flax could look like:

```python
from flax.traverse_util import flatten_dict, unflatten_dict

def merge(*dicts):
    flat_merged = {}
    for d in dicts:
        flat_merged.update(flatten_dict(d))
    return unflatten_dict(flat_merged)
```
It's important that the merge is recursive, to give users flexibility to freeze arbitrary sets of parameters (e.g. only training biases of the layers, or perhaps unfreezing only batchnorm).

This would change the training step's loss function to:
```python
def loss_fn(trainable_params):
  variables = {'params': merge(trainable_params, state.frozen_params), 
               **state.model_state}
  outputs, new_model_state = state.apply_fn(
      variables, inputs, mutable=['batch_stats'])
  loss = xent_loss(outputs, labels)
  return loss, new_model_state
```

This makes transfer learning straightforward, while the normal use cases is not impacted significantly (the user can just use `trainable_params` as usual). But there are some caveats:
- `.params` is unexpectedly expensive for a property, since you have to do a merge upon every access. This doesn't matter inside a JIT'd function, does outside of JIT.
- I'm not sure how FrozenDicts work under the hood, but I would assume you need to do add some freezing and unfreezing to the code I provided to make it run, which would make it more expensive.

Reply:
Another alternative is to keep the `TrainState` as is in the proposal, and let users override and to a `frozen_params`, and also provide a `merge` function through the library (which sounds like what you were already proposing @andsteing).

Reply:
On second thought, @andsteing your suggestion makes more sense from a library perspective (letting the user override to add a `frozen_params` attribute and change the  `.params` themselves if needed). I still think it would be beneficial to provide a [`haiku.data_structures.merge`](https://dm-haiku.readthedocs.io/en/latest/api.html#module-haiku.data_structures) equivalent to facilitate this.

Reply:
@n2cholas I simplified `TrainState` with your suggestions

wrt optimizing only some parameters : this is currently supported via `MultiOptimizer` (i.e. `optax.masked()`) - if we want to make this usecase more efficient for cases where we only need gradients for a fraction of the `params`, then we could do this via multiple collections; note though that AFAIK this is not a common pattern in Flax and that there is some special logic for the `params` collection that would need to be replicated for other `params`-like collections. I think that's an interesting usecase but maybe for another FLIP. Curious to hear @avital's take on this.

Reply:
@andsteing thanks for addressing my comments, that makes sense! I'm curious to hear @avital's thoughts as well.

Reply:
Looks like with the latest `TrainState` proposal, you could achieve the transfer learning use-case by either using the new `masked` in Optax, or re-implementing the `update` method in your "user-space" TrainState subclass to mix parts of the old and new parameter dict (either using hk.data_structure or something similar, or directly). So for my original request and what I think @n2cholas was aluding to, I think everything has a simple solution now.

I prefer this current approach than explicitly creating a distinction between "optimized parameters" and "unoptimizer parameters". What if you want to sometimes optimize different subsets, like after training step 100? Make the user be explicit rather than adding a small number of insufficient hooks.